File: problems-extend-adapt/1901-2000/1946_one-substring-upgrade/solution.cpp
class Solution {
  public:
    string largestAfterMutation(string num, vector<int> &change) {
        // Greedy: the leftmost digit change strictly improves is where the
        // mutation must start -- an earlier digit is more significant, so
        // improving it dominates any later start. Extend through every
        // non-hurting digit (change[d] >= d) and stop at the first hurting
        // one, since the mutated substring must stay contiguous.
        bool started = false;
        for (int i = 0; i < (int)num.size(); ++i) {
            int d = num[i] - '0';
            if (change[d] > d) {
                started = true;
                num[i] = (char)('0' + change[d]);
            } else if (change[d] < d && started) {
                break;
            }
        }
        return num;
    }
};

File: problems-extend-adapt/1901-2000/1946_one-substring-upgrade/solution.py
from typing import List, Optional


class Solution:
    def largestAfterMutation(self, num: str, change: List[int]) -> str:
        # Greedy: the leftmost digit change strictly improves is where the
        # mutation must start -- an earlier digit is more significant, so
        # improving it dominates any later start. Extend through every
        # non-hurting digit (change[d] >= d) and stop at the first hurting
        # one, since the mutated substring must stay contiguous.
        digits = list(num)
        started = False
        for index, character in enumerate(num):
            d = ord(character) - 48
            if change[d] > d:
                started = True
                digits[index] = str(change[d])
            elif change[d] < d and started:
                break
        return "".join(digits)
